src/ctap/storage.rs

```diff
@@ -467,6 +471,70 @@ impl PersistentStore {
 
+    /// Reads the byte vector stored as the serialized large blobs array.
+    ///
+    /// If too few bytes exist at that offset, return the maximum number
+    /// available. This includes cases of offset being beyond the stored array.
+    pub fn get_large_blob_array(
+        &self,
+        mut byte_count: usize,
+        mut offset: usize,
+    ) -> Result<Vec<u8>, Ctap2StatusCode> {
+        if self.store.find(key::LARGE_BLOB_SHARDS.start)?.is_none() {
+            return Ok(vec![
+                0x80, 0x76, 0xBE, 0x8B, 0x52, 0x8D, 0x00, 0x75, 0xF7, 0xAA, 0xE9, 0x8D, 0x6F, 0xA5,
+                0x7A, 0x6D, 0x3C,
+            ]);
+        }
+        let mut output = Vec::with_capacity(byte_count);
+        while byte_count > 0 {
+            let shard_offset = offset % SHARD_SIZE;
+            let shard_length = cmp::min(SHARD_SIZE - shard_offset, byte_count);
+
+            let shard_key = key::LARGE_BLOB_SHARDS.start + offset / SHARD_SIZE;
+            if !key::LARGE_BLOB_SHARDS.contains(&shard_key) {
+                // This request should have been caught at application level.
+                return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+            }
+            let shard_entry = self.store.find(shard_key)?.unwrap_or_default();
+            if shard_entry.len() < shard_offset + shard_length {
+                // If fewer bytes exist than requested, return them all.
+                output.extend(&shard_entry[shard_offset..]);
+                return Ok(output);
+            }
+            output.extend(&shard_entry[shard_offset..shard_offset + shard_length]);
+            offset += shard_length;
+            byte_count -= shard_length;
+        }
+        Ok(output)
+    }
+
+    /// Sets a byte vector as the serialized large blobs array.
+    pub fn commit_large_blob_array(
+        &mut self,
+        large_blob_array: &[u8],
+    ) -> Result<(), Ctap2StatusCode> {
+        if large_blob_array.len() > MAX_LARGE_BLOB_ARRAY_SIZE {
+            return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+        }
+        const MIN_SHARD_KEY: usize = key::LARGE_BLOB_SHARDS.start;
+        const SHARD_COUNT: usize = key::LARGE_BLOB_SHARDS.end - MIN_SHARD_KEY;
+        let mut transactions = Vec::with_capacity(SHARD_COUNT);
+        for shard_key in MIN_SHARD_KEY..key::LARGE_BLOB_SHARDS.end {
+            let large_blob_index = (shard_key - MIN_SHARD_KEY) * SHARD_SIZE;
+            if large_blob_array.len() > large_blob_index {
+                let shard_length = cmp::min(SHARD_SIZE, large_blob_array.len() - large_blob_index);
+                transactions.push(StoreUpdate::Insert {
+                    key: shard_key,
+                    value: &large_blob_array[large_blob_index..large_blob_index + shard_length],
+                });
+            } else {
+                transactions.push(StoreUpdate::Remove { key: shard_key });
+            }
+        }
+        Ok(self.store.transaction(&transactions)?)
```

Comment:
Just as a note (maybe we want to add as a comment in code), but we might want to optimize this function (although it's probably not useful since this should be called very rarely):
1. Only write a shard if it is different from before. This optimization could actually be done in the persistent storage using a flag (not sure about the design yet). This would permit use-cases that are append-only or modify bytes only (not insert or delete) to be much more efficient.
2. Do not add useless remove to the transaction. Each remove is 1 word of lifetime. This is not very useful though compared to the other costs. But it would come for free if the previous step is implemented here (and not in the persistent store).

Reply:
In fact, a `largeBlob` always ends with a 16 byte hash (so far, the storage did not care). If we had the old blob length available, we could check easily. I'm not sure if we need this kind of protection though. In any case, I can add more `find`s to not adding unused `remove`s.

Reply:
If we still want (1), I'd rather implement it one level higher even. But I don't think we have to, I don't expect a platform to write without reading first.
For me (2) is more universally useful, as it helps with store lifetime after a single commit. Therefore (2) is implemented.

Reply:
There's probably a misunderstanding on (1). The platform has no power for this. This is an optimization at the storage level. I don't mean to skip writing the whole blob, just to skip writing a shard. This can have huge benefits in the case mentioned (localized byte updates, append and remove from tail).

I implemented a generic fragmentation module in #270 because this might be useful at other places too and it will keep the CTAP storage simpler. Also, we can more easily fuzz it (not sure it's useful but I still added to my list in #184).

Reply:
Using the new fragment, so should be fixed.